# Analyzing business data

---

# Imports

#### Python libraries

In [1]:
import pandas as pd
pd.set_option('display.max_columns', 500)
# pd.set_option('use_inf_as_na', True)

import numpy as np

import os

import re

import matplotlib.pyplot as plt

from datetime import *

import plotly.graph_objects as go
from plotly.subplots import make_subplots

import pprint as pp

import math

#### Ancillary modules

In [2]:
%load_ext autoreload
%autoreload 2

from BI_params import *
from BI_funcs import *

---

# Data analysis - Sales

## Import and clean data

In [3]:
dfs = pd.read_csv(os.path.join(data_loc, sales_data))

/Users/rp_mbp/Documents/ReposRob_RobPer/Venv_PuntoClinico/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3145: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [4]:
dfs = dfs_clean(dfs)

/Users/rp_mbp/Documents/ReposRob_RobPer/Venv_PuntoClinico/lib/python3.8/site-packages/pandas/core/computation/expressions.py:68: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  return op(a, b)


#### Reviewing specialties results

In [ ]:
dfs["Especialidad_match"].value_counts()

#### Reviewing business line results

In [ ]:
dfs["BusLine_match"].value_counts()

## Analysis 1
- How have sales for each location behaved? (separating laboratory)

### Analysis 1.1
- Weekly basis

In [ ]:
dfsx = data_processing_sales_A1p1(dfs)

In [ ]:
dfsx

In [ ]:
graph_sales_A1(dfsx)

### Analysis 1.2
- Monthly basis

In [ ]:
dfs

---

## Analysis 2
- How have sales for each location behaved? (merging laboratory)

### Analysis 2.1
- Weekly basis

In [ ]:
dfsx = data_processing_sales_A2p1(dfs)

In [ ]:
dfsx

In [ ]:
graph_sales_A2p1(dfsx)

### Analysis 2.2
- Montly basis

#### Base code

In [74]:
dfsx = data_processing_sales_A2p2(dfs)

In [75]:
dfsx

Loc_Hom      BASILICA       COAPA  Location_not_found        MARINA
Date                                                               
2018-10  0.000000e+00        0.00                 0.0  6.184679e+04
2018-11  0.000000e+00        0.00                 0.0  2.641516e+05
2018-12  0.000000e+00        0.00                 0.0  3.655863e+05
2019-1   0.000000e+00        0.00                 0.0  5.055935e+05
2019-2   0.000000e+00        0.00                 0.0  5.626424e+05
2019-3   0.000000e+00        0.00                 0.0  6.842815e+05
2019-4   0.000000e+00        0.00                 0.0  6.644799e+05
2019-5   0.000000e+00        0.00                 0.0  7.499423e+05
2019-6   0.000000e+00        0.00                 0.0  7.344469e+05
2019-7   4.238034e+05        0.00                 0.0  7.758847e+05
2019-8   5.676041e+05        0.00                 0.0  7.934012e+05
2019-9   5.519325e+05        0.00                 0.0  7.598249e+05
2019-10  7.222716e+05        0.00                 0.0  8.386164e+05
2019-11  6.985662e+05        0.00                 0.0  8.045164e+05
2019-12  5.641381e+05        0.00                 0.0  6.992283e+05
2020-1   7.226958e+05        0.00                 0.0  7.977527e+05
2020-2   6.451268e+05    89359.60                 0.0  8.025281e+05
2020-3   7.188750e+05   501584.00                 0.0  8.700294e+05
2020-4   6.893573e+05   503784.05                 0.0  7.267278e+05
2020-5   6.775497e+05   614082.00                 0.0  8.411297e+05
2020-6   8.362951e+05   849299.95                 0.0  1.050106e+06
2020-7   1.072987e+06  1092909.45                 0.0  1.203738e+06
2020-8   1.025888e+06  1147993.25                 0.0  1.079490e+06
2020-9   6.789326e+05   848735.75              8875.0  6.982982e+05

In [76]:
dfsx.drop("Location_not_found", axis=1, inplace=True)

In [77]:
graph_sales_A2p2(dfsx)

#### Further analysis

In [ ]:
dfx = dfsx.copy()

In [ ]:
## Filtering for relevant time period
m1 = dfx.index[list(dfx.index).index("2020-4"):list(dfx.index).index("2020-9")]
dfx = dfx.loc[m1, :]

## Calculating percent change
for col in dfx.columns:
    dfx.loc[:, col +'_pc'] = dfx[col].diff()/dfx[col].shift()*100
    dfx.loc[:, col +'_pc'] = dfx[col +'_pc'].fillna(0)
    dfx.loc[:, col +'_pc'] = dfx[col +'_pc'].replace([np.inf, -np.inf], 0)
    
dfx

---

## Analysis 3
- How have sales for each location separately behaved weekly? (merging laboratory)

In [ ]:
dfsx = data_processing_sales_A3(dfs)

In [ ]:
dfsx

In [ ]:
graph_sales_A3(dfsx)

---

## Analysis 4
- How have sales behaved per location per business line?

#### Base code

In [14]:
dfsx = data_processing_sales_A4(dfs)

In [15]:
dfsx

Loc_Hom             BASILICA                                          \
BusLine_match       CONSULTA    LABORATORIO   OTRA_LN PROCEDIMIENTOS   
BillDate BillDate                                                      
2018     10             0.00       0.000000      0.00            0.0   
         11             0.00       0.000000      0.00            0.0   
         12             0.00       0.000000      0.00            0.0   
2019     1              0.00       0.000000      0.00            0.0   
         2              0.00       0.000000      0.00            0.0   
         3              0.00       0.000000      0.00            0.0   
         4              0.00       0.000000      0.00            0.0   
         5              0.00       0.000000      0.00            0.0   
         6              0.00       0.000000      0.00            0.0   
         7         271028.00   48117.704238   4594.24        74027.5   
         8         358052.40   71704.500000   7155.20       100402.0   
         9         365219.10   59119.600000   8374.00        93810.8   
         10        424633.10   78549.000000  20987.50       145752.0   
         11        399040.10   89056.100000  27293.00       137937.0   
         12        308639.30   77610.750000  20302.00       130046.0   
2020     1         408475.50   97059.500000  24024.80       149716.0   
         2         346435.50   80037.500000  19974.80       156129.0   
         3         389100.00   83406.000000   7914.00       186158.0   
         4         343473.00   92343.300000   8726.00       198545.0   
         5         370085.75   98982.500000   9300.00       157499.0   
         6         418681.60  140393.000000  25644.50       208492.0   
         7         525151.80  206084.500000  18471.50       274991.0   
         8         547798.30  212666.500000  13649.50       201954.0   
         9         342836.60  138912.000000  10964.00       146250.0   

Loc_Hom                            COAPA                                       \
BusLine_match     ULTRASONIDO   CONSULTA LABORATORIO   OTRA_LN PROCEDIMIENTOS   
BillDate BillDate                                                               
2018     10               0.0       0.00         0.0      0.00            0.0   
         11               0.0       0.00         0.0      0.00            0.0   
         12               0.0       0.00         0.0      0.00            0.0   
2019     1                0.0       0.00         0.0      0.00            0.0   
         2                0.0       0.00         0.0      0.00            0.0   
         3                0.0       0.00         0.0      0.00            0.0   
         4                0.0       0.00         0.0      0.00            0.0   
         5                0.0       0.00         0.0      0.00            0.0   
         6                0.0       0.00         0.0      0.00            0.0   
         7            26036.0       0.00         0.0      0.00            0.0   
         8            30290.0       0.00         0.0      0.00            0.0   
         9            25409.0       0.00         0.0      0.00            0.0   
         10           52350.0       0.00         0.0      0.00            0.0   
         11           45240.0       0.00         0.0      0.00            0.0   
         12           27540.0       0.00         0.0      0.00            0.0   
2020     1            43420.0       0.00         0.0      0.00            0.0   
         2            42550.0   62954.50     10675.1   4264.00        11466.0   
         3            52297.0  331128.50     43916.4  12391.60       113847.5   
         4            46270.0  333474.70     38253.6   8789.25       114576.5   
         5            41682.5  392202.00     57591.0   7037.00       140872.0   
         6            43084.0  514831.95    114765.5  21344.00       177293.5   
         7            48288.0  674896.45    146176.5  11645.50       229181.0   
         8            49820.0 

#### Further analysis

In [58]:
dfx = dfsx.copy()

In [59]:
set_loc = "MARINA"

In [60]:
## Selecting row and columns related with the location and period of interest
mr1 = dfx.index.get_level_values(0) == 2020
mr2 = (dfx.index.get_level_values(1) >= 4) & (dfx.index.get_level_values(1) <= 8)
mc1 = dfx.columns.get_level_values(0) == set_loc

dfx = dfx.loc[(mr1 & mr2), mc1]


## Some quick cleanup
#### Removing multi columns names
dfx.columns = dfx.columns.droplevel()
#### Calculating the participation of each business line in total income
dfx["Month_total"] = dfx.sum(axis=1)
for col in [col for col in dfx.columns if "total" not in col]:
    dfx.loc[:, col + "_part"] = dfx[col]/dfx["Month_total"]*100
dfx.drop(["Month_total"], axis=1, inplace=True)

#### Converting multi-index to single-index
dfx["Date"] = dfx.index.get_level_values(0).astype("str") + "-" + dfx.index.get_level_values(1).astype("str")
dfx.set_index("Date", inplace=True, drop=True)


## Adding metrics related with the percent change per business line
for col in [col for col in dfx.columns if "_part" not in col]:
    dfx.loc[:, col + "_pc"] = dfx[col].diff()/dfx[col].shift()
    dfx.loc[:, col +'_pc'] = dfx[col +'_pc'].fillna(0)
    dfx.loc[:, col +'_pc'] = dfx[col +'_pc'].replace([np.inf, -np.inf], 0)

In [61]:
## Filtering results for display
col_filter = "_pc"

In [62]:
dfx.loc[:, [col for col in dfx.columns if col_filter in col]]

BusLine_match  CONSULTA_pc  LABORATORIO_pc  OTRA_LN_pc  PROCEDIMIENTOS_pc  \
Date                                                                        
2020-4            0.000000        0.000000    0.000000           0.000000   
2020-5            0.150855       -0.106179   -0.593006           0.306592   
2020-6            0.213479        0.969553    1.216761           0.097962   
2020-7            0.155794        0.323957   -0.318851           0.062871   
2020-8            0.047387       -0.080798   -0.199891          -0.338777   

BusLine_match  ULTRASONIDO_pc  
Date                           
2020-4               0.000000  
2020-5               0.488258  
2020-6              -0.033336  
2020-7               0.221483  
2020-8              -0.005203

In [63]:
fig = go.Figure()


## Graphing bars
for col in [col for col in dfx.columns if "_pc" not in col and "_part" not in col]:
    fig.add_trace(
        go.Bar(
            x = dfx.index,
            y = dfx[col],
            name = col,
            text = dfx[col].astype("float"),
            textposition = "inside",
            texttemplate = "$%{value:.2s}"
        )
    )


## Labels and format
fig.update_layout(
    #### Labels
    title = str(set_loc).title() + " - Ventas por línea de negocio",
    xaxis_title = "Mes",
    yaxis_title = "Ingresos [$MXN]",
#     yaxis_tickformat = ",.2f",
    yaxis_tickprefix = "$",
#     separators = ".,",
    
    #### Dimensions
    autosize = False,
    width = 1500,
    height = 750
)


## X-axis
fig.update_xaxes(
    dtick="M1"
)

    
fig.show()

---

## Analysis 5
- How have sales behaved per location, per speciality?

#### Base code

In [9]:
dfsx = data_processing_sales_A5(dfs)

In [10]:
dfsx

Loc_Hom               BASILICA                                          \
Especialidad_match ALERGOLOGIA CARDIOLOGIA DERMATOLOGIA ENDOCRINOLOGIA   
BillDate BillDate                                                        
2018     10                0.0         0.0          0.0           0.00   
         11                0.0         0.0          0.0           0.00   
         12                0.0         0.0          0.0           0.00   
2019     1                 0.0         0.0          0.0           0.00   
         2                 0.0         0.0          0.0           0.00   
         3                 0.0         0.0          0.0           0.00   
         4                 0.0         0.0          0.0           0.00   
         5                 0.0         0.0          0.0           0.00   
         6                 0.0         0.0          0.0           0.00   
         7                 0.0         0.0     170961.5        4092.60   
         8                 0.0         0.0     197682.1        5026.70   
         9             25689.0         0.0     182489.5        3547.40   
         10            38318.0         0.0     227795.7        2547.00   
         11            31598.4         0.0     242507.1        3545.90   
         12            44651.0         0.0     194043.7        4116.75   
2020     1             12684.0         0.0     221824.7        5040.20   
         2               873.0         0.0     199989.0        7118.50   
         3                 0.0         0.0     208732.6       10064.00   
         4                 0.0         0.0     214775.5       41884.00   
         5                 0.0         0.0     200212.5       48833.25   
         6                 0.0         0.0     213449.1       73635.00   
         7                 0.0         0.0     266223.5      128716.10   
         8                 0.0         0.0     297095.3      132070.50   
         9                 0.0         0.0     197478.0      105029.00   

Loc_Hom                                                               \
Especialidad_match GASTROENTEROLGIA GASTROENTEROLOGIA    GINECOLOGIA   
BillDate BillDate                                                      
2018     10                     0.0               0.0       0.000000   
         11                     0.0               0.0       0.000000   
         12                     0.0               0.0       0.000000   
2019     1                      0.0               0.0       0.000000   
         2                      0.0               0.0       0.000000   
         3                      0.0               0.0       0.000000   
         4                      0.0               0.0       0.000000   
         5                      0.0               0.0       0.000000   
         6                      0.0               0.0       0.000000   
         7                      0.0           90139.0   64904.596238   
         8                      0.0          112871.8   73146.200000   
         9                      0.0           93350.0   64679.100000   
         10                     0.0          123151.6  130901.500000   
         11                     0.0          126116.2  142346.200000   
         12                     0.0           87779.5  103758.000000   
2020     1                      0.0          115963.0  162601.000000   
         2                      0.0          101920.7  151366.000000   
         3                      0.0          112194.1  187825.000000   
         4                      0.0           78971.5  140847.000000   
         5                      0.0          115457.0  136023.500000   
         6                      0.0          119734.5  193195.000000   
         7                      0.0          130554.2  267816.000000   
         8                      0.0          129117.0  233173.500000   
         9                      0.0           85751.1  148141.000000   

Loc_Hom                                                 

#### Further analysis

In [11]:
dfx = dfsx.copy()

In [12]:
## Select a specific location
sel_loc = "BASILICA"

In [13]:
## Filter columns and rows according to analysis
mc1 = dfx.columns.get_level_values(0) == sel_loc
mr1 = (dfx.index.get_level_values(0) == 2020) & ((dfx.index.get_level_values(1) >= 4) & (dfx.index.get_level_values(1) <= 8))
dfx = dfx.loc[mr1, mc1]
dfx.columns = dfx.columns.droplevel()

## Changing index to one column string
dfx["Date"] = dfx.index.get_level_values(0).astype("str") + "-" + dfx.index.get_level_values(1).astype("str")
dfx.set_index("Date", inplace=True, drop=True)

dfx

Especialidad_match  ALERGOLOGIA  CARDIOLOGIA  DERMATOLOGIA  ENDOCRINOLOGIA  \
Date                                                                         
2020-4                      0.0          0.0      214775.5        41884.00   
2020-5                      0.0          0.0      200212.5        48833.25   
2020-6                      0.0          0.0      213449.1        73635.00   
2020-7                      0.0          0.0      266223.5       128716.10   
2020-8                      0.0          0.0      297095.3       132070.50   

Especialidad_match  GASTROENTEROLGIA  GASTROENTEROLOGIA  GINECOLOGIA  \
Date                                                                   
2020-4                           0.0            78971.5     140847.0   
2020-5                           0.0           115457.0     136023.5   
2020-6                           0.0           119734.5     193195.0   
2020-7                           0.0           130554.2     267816.0   
2020-8                           0.0           129117.0     233173.5   

Especialidad_match  MEDICINA GENERAL  NEFROLOGIA  NEUMOLOGIA  NEUROLOGIA  \
Date                                                                       
2020-4                       10467.0         0.0         0.0         0.0   
2020-5                        9502.0         0.0         0.0         0.0   
2020-6                        9414.0         0.0         0.0         0.0   
2020-7                       14501.0         0.0         0.0         0.0   
2020-8                       11841.0         0.0         0.0         0.0   

Especialidad_match  NUTRICION  OTORRINOLARINGOLOGIA  OTRA_ESP  PEDIATRIA  \
Date                                                                       
2020-4                    0.0                   0.0   54006.3        0.0   
2020-5                    0.0                   0.0   31593.5        0.0   
2020-6                 4630.5                   0.0   26867.0        0.0   
2020-7                14116.0                   0.0   46110.0        0.0   
2020-8                17446.0                   0.0   26509.0        0.0   

Especialidad_match  PROCTOLOGIA  UROLOGIA  
Date                                       
2020-4                      0.0  148406.0  
2020-5                      0.0  135928.0  
2020-6                      0.0  195370.0  
2020-7                      0.0  204950.0  
2020-8                      0.0  178636.0

In [62]:
def top_5_columns(row, no_tops=2):
    
    
    ## Key attributes 
    row_total = row.sum()
    sales_no_top = round(row_total - sum(row.nlargest(no_tops).values), 1)
    
    
    ## Obtaining top specialities
    res = {}
    for top in range(1, no_tops + 1):
        res[str(top) + "_Top"] = {
            "Spec": row.nlargest(top).index[-1],
            "Sales": row.nlargest(top).values[-1],
            "Part": round((row.nlargest(top).values[-1]/row_total)*100, 1)
        }
    
    
    ## Adding to dict all specialities that were not included in the top
    res["Remaining"] = {
        "Spec": "Remaining",
        "Sales": sales_no_top,
        "Part": round((sales_no_top/row_total)*100, 1)
    }
    
    
    return res

In [63]:
## Creating series with a selection of top 5 specialities and a grouping of remainings
res = dfx.apply(lambda x: top_5_columns(x, 5), axis=1)

In [64]:
## Craeting nested dictionary with results to obtain dataframe
i = 0
resf = {}
for x in res:
    for y in x:
        for z in x[y]:
            resf[(res.index[i], z, y)] = x[y][z]
    i += 1

# resf

In [65]:
## Creting dataframe with results
mux = pd.MultiIndex.from_tuples(resf.keys())
dfx = pd.DataFrame(resf.values(), index=mux).unstack()
dfx.columns = dfx.columns.droplevel()

In [66]:
dfx

1_Top         2_Top        3_Top              4_Top  \
2020-4 Part           31.2          21.5         20.4               11.5   
       Sales        214776        148406       140847            78971.5   
       Spec   DERMATOLOGIA      UROLOGIA  GINECOLOGIA  GASTROENTEROLOGIA   
2020-5 Part           29.5          20.1         20.1                 17   
       Sales        200212        136024       135928             115457   
       Spec   DERMATOLOGIA   GINECOLOGIA     UROLOGIA  GASTROENTEROLOGIA   
2020-6 Part           25.5          23.4         23.1               14.3   
       Sales        213449        195370       193195             119734   
       Spec   DERMATOLOGIA      UROLOGIA  GINECOLOGIA  GASTROENTEROLOGIA   
2020-7 Part             25          24.8         19.1               12.2   
       Sales        267816        266224       204950             130554   
       Spec    GINECOLOGIA  DERMATOLOGIA     UROLOGIA  GASTROENTEROLOGIA   
2020-8 Part             29          22.7         17.4               12.9   
       Sales        297095        233174       178636             132070   
       Spec   DERMATOLOGIA   GINECOLOGIA     UROLOGIA     ENDOCRINOLOGIA   

                          5_Top  Remaining  
2020-4 Part                 7.8        7.6  
       Sales            54006.3      52351  
       Spec            OTRA_ESP  Remaining  
2020-5 Part                 7.2        6.1  
       Sales            48833.2    41095.5  
       Spec      ENDOCRINOLOGIA  Remaining  
2020-6 Part                 8.8        4.9  
       Sales              73635    40911.5  
       Spec      ENDOCRINOLOGIA  Remaining  
2020-7 Part                  12          7  
       Sales             128716      74727  
       Spec      ENDOCRINOLOGIA  Remaining  
2020-8 Part                12.6        5.4  
       Sales             129117      55796  
       Spec   GASTROENTEROLOGIA  Remaining

## Analysis 6
- How have sales behaved per location, per business line, per speciality?

#### Base code

In [ ]:
dfsx = data_processing_sales_A6(dfs)

In [ ]:
dfsx

#### Further analysis

---

## Analysis 7
- How have sales behaved per location, per provider, per business line?

In [ ]:
dfsx = data_processing_sales_A7(dfs)

In [ ]:
dfsx

---

# Data analysis - Appointments

## Importing and cleaning data

In [ ]:
dfa = pd.read_csv(os.path.join(data_loc, appointment_data))

In [ ]:
dfa = dfa_clean(dfa)
dfa

## Analysis 1
- How has the number of confirmed appointments behaved weekly since 2019 for each of the most relevant specialities

In [ ]:
loc='ALL'

In [ ]:
dfax = data_processing_appointments_A1(dfa, loc=loc)

In [ ]:
dfax

In [ ]:
graph_appointments_A1(dfax, loc=loc)

## Analysis 2
- How has the number of confirmed appointments increased quarterly since 2019?

In [ ]:
loc='ALL'

In [ ]:
dfax = data_processing_appointments_A2(dfa, loc=loc)

In [ ]:
dfax

In [ ]:
dfax

In [ ]:
graph_appointments_A2(dfax, loc=loc)

---

# Data analysis - Tickets

In [26]:
df_sapi = pd.read_csv("Data/Emitidos-Punto_Clinico_SAPI-2020-08.csv")
df_sc = pd.read_csv("Data/Emitidos-Punto_Clinico_SC-2020-08.csv")

In [21]:
df_sapi

XML    Rfc Emisor  \
0    1F5F0551-1247-458C-AC22-BCE0B5FEBBD0.xml  PCV180730UJ1   
1    8BFD38DE-5B16-4A34-B199-97714D68E111.xml  PCV180730UJ1   
2    06458633-7A54-462E-AB18-D0EB18773027.xml  PCV180730UJ1   
3    BC1172F5-D755-45F8-AFA7-FCEBA8E704D9.xml  PCV180730UJ1   
4    46EB9572-A7F4-48F6-9D18-7F1C2B67C4D4.xml  PCV180730UJ1   
..                                        ...           ...   
195  50046596-a28d-411d-a385-373e004eda6e.xml  PCV180730UJ1   
196  928b53a6-5fed-4001-a217-9bb3dbea27c7.xml  PCV180730UJ1   
197  58ee7419-a175-4d75-9198-a931609c252e.xml  PCV180730UJ1   
198  1d236376-041d-4ad6-9095-33622e906388.xml  PCV180730UJ1   
199  de7dd268-9ed4-48cf-af87-54f200f8f352.xml  PCV180730UJ1   

                           Nombre Emisor   Rfc Receptor  \
0    PUNTO CLINICO VISIONARIO SAPI DE CV  ROTD9301098R2   
1    PUNTO CLINICO VISIONARIO SAPI DE CV  COOE810712LZ3   
2    PUNTO CLINICO VISIONARIO SAPI DE CV   LMC741212JN6   
3    PUNTO CLINICO VISIONARIO SAPI DE CV  OISG760103JE9   
4    PUNTO CLINICO VISIONARIO SAPI DE CV  HEFI770326LE5   
..                                   ...            ...   
195  PUNTO CLINICO VISIONARIO SAPI DE CV  GAGB710226472   
196  PUNTO CLINICO VISIONARIO SAPI DE CV  GAPA900601451   
197  PUNTO CLINICO VISIONARIO SAPI DE CV  HEGR6908316W0   
198  PUNTO CLINICO VISIONARIO SAPI DE CV  FAMJ7608234V8   
199  PUNTO CLINICO VISIONARIO SAPI DE CV  TEPR731105I94   

                            Nombre Receptor     Tipo      Serie  Folio  \
0                   DENNIS RODRIGUEZ TORRES  ingreso  CAFETALES   3651   
1          EDGAR DIONISIO COVARRUBIOS ORTIZ  ingreso      VILLA   3652   
2    GRUPO DIAGNOSTICO MEDICO PROA SA DE CV  ingreso     MARINA   3653   
3         GLORIA GUADALUPE OLIVARES SANCHEZ  ingreso  CAFETALES   3654   
4                   IVONNE HERNANDEZ FLORES  ingreso      VILLA   3655   
..                                      ...      ...        ...    ...   
195            BLANCA LIDIA GARCIA GONZALEZ  ingreso      VILLA   3818   
196                      AXEL GARCIA POISOT  ingreso      VILLA   3819   
197                REINA HERNANDEZ GALLEGOS  ingreso      VILLA   3820   
198                   JACOBO FRANQUIZ MUÑOZ  ingreso      VILLA   3821   
199                      RUBEN TELLEZ PEREZ  ingreso      VILLA   3822   

                   Fecha  Sub Total  Descuento  Total impuesto Trasladado  \
0    2020-08-01T09:24:55     699.00        0.0                       0.00   
1    2020-08-01T12:49:41    2430.00        0.0                       0.00   
2    2020-08-03T09:06:39  325000.00        0.0                   52000.00   
3    2020-08-03T09:14:37     399.00        0.0                       0.00   
4    2020-08-03T09:31:27     370.69        0.0                      59.31   
..                   ...        ...        ...                        ...   
195  2020-08-31T13:33:27    1300.00        0.0                       0.00   
196  2020-08-31T16:14:22     549.00        0.0                       0.00   
197  2020-08-31T17:16:51     399.00        0.0                       0.00   
198  2020-08-31T18:10:50     399.00        0.0                       0.00   
199  2020-08-31T19:01:03     771.00        0.0                       0.00   

    Nombre Impuesto  Total impuesto Retenido Nombre Impuesto.1     Total  \
0               NaN                      0.0               NaN     699.0   
1               NaN                      0.0               NaN    2430.0   
2         002 - IVA                      0.0               NaN  377000.0   
3               NaN                      0.0               NaN     399.0   
4         002 - IVA                      0.0               NaN     430.0   
..              ...                      ...               ...       ...   
195             NaN                      0.0               NaN    1300.0   
196             NaN                      0.0               NaN     549.0   
197             NaN                      0.0               NaN     399.0   
19

In [32]:
df_sapi["Estado"].unique()

array(['Vigente'], dtype=object)

In [27]:
df_sapi.shape[0]

200

In [30]:
1 - (df_sapi.shape[0] / (df_sapi.shape[0] + df_sc.shape[0]))

0.7014925373134329

In [29]:
df_sc

XML    Rfc Emisor  \
0    e1870129-d4a4-41e9-b656-1efac1010b4e.xml  SMP190128NHA   
1    d01e6b8f-8bf4-4780-b78c-2d14936f4a96.xml  SMP190128NHA   
2    27a6a430-2268-4aca-9c46-ea6e34a91358.xml  SMP190128NHA   
3    e2a8a9f2-16bc-457a-9bd2-c1a48fad71a5.xml  SMP190128NHA   
4    e694c160-6360-4db3-aa98-3866ac2e23a2.xml  SMP190128NHA   
..                                        ...           ...   
465  c0a0eb8b-502b-41d7-af7a-ea94cd546930.xml  SMP190128NHA   
466  e0843bfd-e086-4cee-ae31-048d5ab0f303.xml  SMP190128NHA   
467  f204c401-25d1-402b-bd84-3358970306aa.xml  SMP190128NHA   
468  0eb557fb-4cd7-42d7-b38a-f282eb67c660.xml  SMP190128NHA   
469  3acdb888-3f21-4e0a-81b4-ad475f205eff.xml  SMP190128NHA   

                          Nombre Emisor   Rfc Receptor  \
0    SERVICIOS MEDICOS PUNTO CLINICO SC  CASL660621SXA   
1    SERVICIOS MEDICOS PUNTO CLINICO SC  ROPE800823FE4   
2    SERVICIOS MEDICOS PUNTO CLINICO SC  ROMH6705141F6   
3    SERVICIOS MEDICOS PUNTO CLINICO SC  GAHA9207028B6   
4    SERVICIOS MEDICOS PUNTO CLINICO SC  SOBA760901JS6   
..                                  ...            ...   
465  SERVICIOS MEDICOS PUNTO CLINICO SC   MGB051006HT8   
466  SERVICIOS MEDICOS PUNTO CLINICO SC  GOAJ750428CG0   
467  SERVICIOS MEDICOS PUNTO CLINICO SC  TEPR731105I94   
468  SERVICIOS MEDICOS PUNTO CLINICO SC  SOVL600421HU8   
469  SERVICIOS MEDICOS PUNTO CLINICO SC  SOVL600421HU8   

                                       Nombre Receptor     Tipo      Serie  \
0                                LUISA CHAVARRIA SALAS  ingreso  CAFETALES   
1                           EDGAR DAVID RODRIGUEZ PEÑA  ingreso  CAFETALES   
2                                    HECTOR ROSAS MORA  ingreso  CAFETALES   
3                        ANA GABRIELA GARCIA HERNANDEZ  ingreso      VILLA   
4                             ALMA LUCERO SOSA BLANCAS  ingreso  CAFETALES   
..                                                 ...      ...        ...   
465  MULTISISTEMAS GUJADI BOSS  SEGURIDAD PRIVADA S...  ingreso     MARINA   
466                              JACOB GONZALEZ ALONSO  ingreso  CAFETALES   
467                                 RUBEN TELLEZ PEREZ  ingreso      VILLA   
468                                 LIDIA SOTO VAZQUEZ  ingreso     MARINA   
469                                 LIDIA SOTO VAZQUEZ  ingreso  CAFETALES   

     Folio                Fecha  Sub Total  Descuento  \
0     1306  2020-08-01T08:34:04      399.0        0.0   
1     1307  2020-08-01T08:41:35      399.0        0.0   
2     1308  2020-08-01T09:02:47      399.0        0.0   
3     1309  2020-08-01T10:12:16      899.0        0.0   
4     1310  2020-08-01T10:20:34      599.0        0.0   
..     ...                  ...        ...        ...   
465   1770  2020-08-31T17:04:30      399.0        0.0   
466   1771  2020-08-31T18:47:44     1755.0        0.0   
467   1772  2020-08-31T18:59:25      399.0        0.0   
468   1773  2020-08-31T19:26:13      150.0        0.0   
469   1774  2020-08-31T19:33:35      158.0        0.0   

     Total impuesto Trasladado  Nombre Impuesto  Total impuesto Retenido  \
0                            0              NaN                      0.0   
1                            0              NaN                      0.0   
2                            0              NaN                      0.0   
3                            0              NaN                      0.0   
4                            0              NaN                      0.0   
..                         ...              ...                      ...   
465                          0              NaN                      0.0   
466                          0              NaN                      0.0   
467                          0              NaN                      0.0   
468                          0              NaN                      0.0   
469                          0              NaN                      0.0   

    Nombre Impuesto.1   Total                        

In [34]:
df_sc["Estado"].value_countscounts()

Vigente      468
Cancelado      2
Name: Estado, dtype: int64

In [25]:
df_sc.shape[0]

470

#### Información acerca de las facturas
**- Número de facturas generadas (total):** 670 (100%)
    **- Número de facturas generadas (mes agosto) (SAPI):** 200 (29.85%)
    **- Número de facturas generadas (mes agosto) (SC):** 470 (70.15%)
    
- Número de facturas canceladas (total): 2 (100%)
    - Número de facturas canceladas (SAPI): 0 (0%)
    - Número de facturas canceladas (SC): 2 (100%)

# *Notes*

#### Substituting characters with re

In [ ]:
x = 'Melón'

In [ ]:
x = re.sub('ó', 'o', x)

In [ ]:
x

#### Creatting formatted tables related with parameters

**Doctors dicctionary**

In [78]:
meds_dict_ref

{'ARROYO COUTIÑO, IBER': {'Especialidad': ['UROLOGIA'],
  'Sitios': ['Punto Clínico- Coapa']},
 'ASCENCIO BARRIENTOS, CARLOS NORBERTO': {'Especialidad': ['GASTROENTEROLOGIA'],
  'Sitios': ['Punto Clinico- Marina',
   'Punto Clínico- Basílica',
   'LABORATORIO- Basílica']},
 'CHAYA SALGADO, SAID': {'Especialidad': ['DERMATOLOGIA'],
  'Sitios': ['Punto Clinico- Marina', 'LABORATORIO- Marina']},
 'CITALAN PROJ, HARRY WILLIAM': {'Especialidad': ['GASTROENTEROLOGIA'],
  'Sitios': ['Punto Clínico- Coapa']},
 'ENRIQUEZ ROMAN, ARLETTE': {'Especialidad': ['ENDOCRINOLOGIA'],
  'Sitios': ['Punto Clínico- Basílica',
   'Punto Clinico- Marina',
   'LABORATORIO- Marina',
   'LABORATORIO- Basílica']},
 'ERAZO PEREZ, LUSVI LUDGARDIZ': {'Especialidad': ['NEUMOLOGIA'],
  'Sitios': ['Punto Clínico- Coapa', 'LABORATORIO- Coapa']},
 'FRANCO ESTRADA, SILVIA': {'Especialidad': ['GASTROENTEROLOGIA'],
  'Sitios': ['Punto Clinico- Marina', 'Punto Clínico- Basílica']},
 'GARCIA CRUZ, LILIANA ESTHER': {'Especiali

In [86]:
pd.DataFrame.from_dict(meds_dict_ref, orient="index").to_csv("Doctors.csv")

**Specialities reference**

In [87]:
specialties_ref

{'GINECOLOGIA': ['GINECOLOGIA',
  'ENDOVAGINAL',
  'TRANSVAGINAL',
  'GINECO',
  'CERVIX',
  'VAGIN',
  'DIU',
  'OBSTETRIC',
  'EMBARAZ',
  'PROLACTIN',
  'COLPOSCOPIA',
  'PAPANICOLAOU'],
 'DERMATOLOGIA': ['DERMATOLOGIA'],
 'UROLOGIA': ['UROLOGIA', 'PROST', 'RENAL', 'PENESCOPIA', 'ANDROSCOPIA'],
 'PEDIATRIA': ['PEDIATRIA'],
 'ENDOCRINOLOGIA': ['ENDOCRINOLOGIA', 'TIROIDEO', 'TIROIDES'],
 'OTORRINOLARINGOLOGIA': ['OTORRINOLARINGOLOGIA',
  'OTORRINOLARINGO',
  'CERUMEN'],
 'GASTROENTEROLOGIA': ['GASTROENTEROLOGIA', 'GASTROENTEROLGIA'],
 'MEDICINA GENERAL': ['MEDICINA GENERAL'],
 'NEUMOLOGIA': ['NEUMOLOGIA'],
 'CARDIOLOGIA': ['CARDIOLOGIA'],
 'NUTRICION': ['NUTRICION'],
 'ALERGOLOGIA': ['ALERGOLOGIA'],
 'NEUROLOGIA': ['NEUROLOG'],
 'PROCTOLOGIA': ['PROCTOLOG']}

In [97]:
pd.DataFrame(dict([(key, pd.Series(val)) for key, val in specialties_ref.items()])).fillna("-").T.to_csv("spec_refs.csv")

**Business lines**

In [98]:
business_line_ref

{'CONSULTA': ['CONSULTA MEDICA',
  'CONSULTA ESPECIALISTA',
  'CONSULTA DE',
  'CONSULTA MEDICINA GENERAL',
  'CONSULTA GENERAL',
  'OTORRINOLARINGOLOGIA',
  'CARDIOLOGIA',
  'TELECONSULTA',
  'DIETA',
  'REVISION'],
 'LABORATORIO': ['CHECK UP',
  'PERFIL TIROIDEO',
  'BIOMETRIA HEMATICA',
  'QUIMICA DE',
  'HEMOGLOBINA',
  'INSULINA',
  'CULTIVO',
  'EXAMEN',
  'PRUEBA',
  'COPROCULTIVO',
  'AMILASA',
  'VDRL',
  'AMIBA',
  'HORMONA',
  'WINTROBE',
  'GLUCOSA',
  'PROLACTINA',
  'TIEMPO DE',
  'PERFIL',
  'GRUPO SANGUINEO',
  'VIRUS INMUNODEFICIENCIA',
  'HIV',
  'CITOLOGIA',
  'ANTICUERPOS',
  'ANTIGENO',
  'PROGESTERONA',
  'DEHIDROEPIANDROSTERONA',
  'TRANSAMINASA'],
 'ULTRASONIDO': ['ULTRASON'],
 'PROCEDIMIENTOS': ['ELECTROFULGURACION',
  'COLPOSCOPIA',
  'PAPANICOLAOU',
  'RASTREO',
  'INFOGRAFIA',
  'BIOPSIA',
  'BIOPISA',
  'APLICACION DE',
  'TOMA',
  'RETIRO',
  'EXTRACCCION',
  'INSERCION',
  'ENVIADO A',
  'CRIOTERAPIA',
  'SCOPIA',
  'CIRUGIA',
  'INYECCION',
  'ELECTROCAR

In [100]:
pd.DataFrame(dict([(key, pd.Series(val)) for key, val in business_line_ref.items()])).fillna("-").T.to_csv("BL_refs.csv")

---
---